web3.js/src/message/index.ts

```diff
@@ -15,18 +19,27 @@ export type MessageHeader = {
   numReadonlyUnsignedAccounts: number;
 };
 
+/**
+ * An address table lookup used to load additional accounts
+ */
+export type MessageAddressTableLookup = {
+  accountKey: PublicKey;
+  writableIndexes: Array<number>;
+  readonlyIndexes: Array<number>;
```

Comment:
I was looking for a way that you could architect this type such that it was impossible to screw it up. As written, you _could_ create malformed data like this:

```typescript
const matl: MessageAddressTableLookup = {
  accountKey: /* ... */,
  writableIndexes: [0, 1],
  readonlyIndexes: [1, 2],
};
```

If the storage was, instead, implemented as a sparse array, it would be impossible to overlap a readable/writeable index.

```typescript
type Readable = 0;  // or 'r'
type Writeable = 1;  // or 'w'
type IndexesWithWriteability = (Readable | Writeable | undefined)[];
const indexes: IndexesWithWriteability = [];
indexes[0] = 1;
indexes[1] = 1;
indexes[1] = 0;  // OVERWRITTEN! Note how overlapping index 1 is impossible here.
indexes[2] = 0;
```

Then you'd do stuff like this at the point of serialization:

```typescript
const readonlyIndexes = [];
const writeableIndexes = [];
indexes.forEach((writeability, index) => {
  if (writeability === 0) {
    readonlyIndexes.push(index);
  } else if (writeability === 1) {
    writeableIndexes.push(index);
  }
});
const writeableIndexesLength = shortvec.encodeLength(
  encodedWriteableIndexesLength,
  writeableIndexes.length,
);
```

Reply:
Oh yeah great point, it's definitely possible to create invalid transaction messages with the `MessageV0` constructor. This `MessageAddressTableLookup` type directly maps to the serialization format of v0 messages (`MessageV0` is just a barebones data class) so it's not meant to be created or handled directly by devs. In this PR you can see that the construction of v0 messages has a nice API which will build these lookup data structures internally in a valid way:
https://github.com/solana-labs/solana/blob/25d25662aa170ee48cb1043cdbc9175c5b178789/web3.js/src/message/v0.ts#L71-L80

https://github.com/solana-labs/solana/blob/25d25662aa170ee48cb1043cdbc9175c5b178789/web3.js/src/message/compiled-keys.ts#L104-L123

Reply:
Yeah, only thing is that `MessageV0` is exported in #27213, so by Murphy's Law someone will _definitely_ eventually create one with hand rolled indices that overlap.

Reply:
I think that's fine. The same holds for any other field in `MessageV0`... `staticAccountKeys` could have dup keys, `instructions` could reference invalid account indexes, the `header` values could be contradictory, etc.